consumergroup.go

```diff
@@ -169,27 +190,32 @@ func (cgi *ConsumergroupInstance) Registered() (bool, error) {
 	return cgi.cg.kz.exists(node)
 }
 
-// Register registers the consumergroup instance in Zookeeper.
-func (cgi *ConsumergroupInstance) Register(topics []string) error {
-	if exists, err := cgi.Registered(); err != nil {
-		return err
-	} else if exists {
-		return ErrInstanceAlreadyRegistered
+// Registered returns current registration of the consumer group instance.
+func (cgi *ConsumergroupInstance) Registration() (*Registration, error) {
+	node := fmt.Sprintf("%s/consumers/%s/ids/%s", cgi.cg.kz.conf.Chroot, cgi.cg.Name, cgi.ID)
+	val, _, err := cgi.cg.kz.conn.Get(node)
+	if err != nil {
+		return nil, err
 	}
 
-	// Create a JSON record that is compatible with the JVM consumer.
-	// I am not quite sure what this information is used for.
+	reg := &Registration{}
+	if err := json.Unmarshal(val, reg); err != nil {
+		return nil, err
+	}
+	return reg, nil
+}
 
+// Register registers the consumergroup instance in Zookeeper.
+func (cgi *ConsumergroupInstance) Register(topics []string) error {
 	subscription := make(map[string]int)
 	for _, topic := range topics {
 		subscription[topic] = 1
 	}
-
-	data, err := json.Marshal(map[string]interface{}{
-		"pattern":      "white_list",
-		"subscription": subscription,
-		"timestamp":    fmt.Sprintf("%d", time.Now().Unix()),
-		"version":      1,
+	data, err := json.Marshal(&Registration{
+		Pattern:      RegPatternStatic,
+		Subscription: subscription,
+		Timestamp:    time.Now().Unix(),
+		Version:      RegDefaultVersion,
 	})
```

Comment:
I overrides the subscription. This way the end result is explicit.

Ever since this proposal I have been browsing the Kafka documentation and source code trying to understand if the allow resubscription. And I have not found explicit mentioning of that anywhere. What I did found is a [list of watches](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/consumer/ZookeeperConsumerConnector.scala#L949-L957) Kafka consumer sets up. To be notified of subscription changes the consumer has to set data watches on all group instances, but it does not do that. So to make the updated subscription visible a client has to deregister and register again.

Given that my changes won't hurt but it might make sense to return an error on re-registration just to encourage the user to do deregister first.